### DISCUSSION TITLE
Modbus Zähler - SAE liefert bei Ganzzahl nur 0 an den SHM

### DISCUSSION THREAD

**Discussion by HansESA** [camueller/SmartApplianceEnabler#501] [Installation und Betrieb]
Hallo,
ich möchte den Verbrauch meiner Lambda Wärmepumpe im SHM darstellen.
Die Wärmepumpe habe ich als Modbus-Zähler mit Leistung, Typ Integer und "immer eingeschaltet" eingerichtet. 
Im Log wird die Leistung in Watt auch richtig angezeigt, aber es wird immer power:0 gesendet.
Sobald ich Integer als Float einstelle, wird eine Leistung >0 gesendet. Allerdings stimmt dieser Wert dann nicht, da ich zwei Datenwörter angeben muss.

Was muss ich anders einstellen?

2023-03-19 15:05:15,282 DEBUG [Timer-0] d.a.s.m.e.ReadHoldingRegisterExecutor [ReadHoldingRegisterExecutor.java:52] F-19101913-000000000001-00: Holding register=1012 value=[1418]
2023-03-19 15:05:15,303 TRACE [pool-80-thread-1] d.a.s.m.MqttClient [MqttClient.java:265] F-19101913-000000000001-00-MQTT-ModbusElectricityMeter: Publish message: topic=sae/F-19101913-000000000001-00/Meter payload={"energy":0.0,"power":0,"time":"2023-03-19T15:05:15.247169","type":"MeterMessage"}

![grafik](https://user-images.githubusercontent.com/128302792/226181742-d64cf9c4-364f-4fd6-89ec-d7b9a858dd64.png)

![grafik](https://user-images.githubusercontent.com/128302792/226181802-43b4afff-454e-4e67-a2c2-7f386089f20e.png)

![grafik](https://user-images.githubusercontent.com/128302792/226181853-d2890d30-2b57-48fd-81c5-9c51b49a93db.png)

**Comment by camueller** [comment]
> Im Log wird die Leistung in Watt auch richtig angezeigt

Woran machst Du das fest? Ich kann das nicht bestätigen auf Basis der Log-Zielen, die  Du gepostet hast.

> Sobald ich Integer als Float einstelle, wird eine Leistung >0 gesendet. Allerdings stimmt dieser Wert dann nicht, da ich zwei Datenwörter angeben muss.

Was willst Du damit sagen? Du kannst auf bei Float die Anzahl der Datenwörter konfigurieren. Außerdem musst Du möglicherweise eine andere Byte-Reihenfolge einstellen. Das sollte aber in der Anleitung Deiner WP stehen. Ansonsten musst Du es ausprobieren. Das habe ich alles in der Doku beschrieben: https://github.com/camueller/SmartApplianceEnabler/blob/master/doc/Modbus_DE.md

**Comment by HansESA** [comment]
Vielen Dank für die schnelle Rückmeldung.
Die Leistungsaufnahme der WP kann ich direkt an deren Display ablesen. Die Angabe im Log value=[1418] bedeutet eine momentane Leistungsaufnahme von 1418 Watt. 

In anderen Diskussionsbeiträgen hatte ich den Hinweis auf Integer als Float gelesen. Dies funktioniert ja aber nur ab 2 Datenwörtern. Allerdings besteht die Leistungsaufnahme nur aus 1 Datenwort (Reg 1012) und das nächste Datenwort (Reg 1013) bereits der COP, was dann zu einem unplausiblen Wert führt. Dieser wird dann aber auch an den SHM gesendet.

![grafik](https://user-images.githubusercontent.com/128302792/226205636-815f8e47-d9bc-4258-af1c-2c4be2e06988.png)

**Reply by camueller** [reply]
Du bekommst einen INT16 aus dem Register und die Leistung ist auch im SAE ein Ganzzahlwert. Also muss nichts konvertiert werden, wie bei "Integer als Float". INT16 bedeutet auch, dass es nur ein Datenwort ist. Eigentlich sollte das mit Wert-Type "Integer" und Datenwörter=1 gehen, wie Du oben in den Screenshots konfiguriert hast. Ich prüfe das mal ...

**Comment by camueller** [comment]
Eigentlich wollte ich heute die 2.2.2 releasen, aber ich konnte den Fehler nachstellen, der dazu führt, dass Du keine Leistung auslesen konntest. Kannst Du bitte testen, ob Du mit der nachfolgend verlinkten Test-Version die Leistung ausgelesen bekommst? Die Zählereinstellung sollten dabei wie ganz oben im Screenshot sein.

https://drive.google.com/file/d/1CTPmBPyiuFIWgtE4dPL0tylqsKD1wWFg/view?usp=share_link

Die Standardwerte musst man übrigens nicht ausfüllen, d.h. die ausgegraute "1" beim Umrechnungsfaktor musst Du nicht eingeben - sie gilt automatisch.

**Comment by HansESA** [comment]
Herzlichen Dank! Mit dem Update funktioniert es jetzt wie gedacht.
Ausschnitt aus dem SunnyPortal:
![grafik](https://user-images.githubusercontent.com/128302792/226417385-4e937358-0752-4314-8e74-311f1c755d5e.png)
